go/vt/vtgateproxy/discovery.go

```diff
@@ -48,225 +41,312 @@ var (
-		return nil, nil, err
+	// Grab the first N addresses, and voila!
+	var addrs []resolver.Address
+	targets = targets[:min(*numConnections, len(targets))]
+	for _, target := range targets {
+		addrs = append(addrs, resolver.Address{Addr: target.addr})
 	}
-	sum := h.Sum(nil)
 
-	log.V(100).Infof("resolved %s to hosts %s addrs: 0x%x, %v\n", r.target.URL.String(), jsonDump(hosts), sum, addrs)
+	log.V(100).Infof("updated targets for %s to %v", r.target.URL.String(), targets)
 
-	return &addrs, sum, nil
+	r.clientConn.UpdateState(resolver.State{Addresses: addrs})
 }
 
-func (r *JSONGateConfigResolver) start() {
-	log.V(100).Infof("Starting discovery checker\n")
-	r.rand = rand.New(rand.NewSource(time.Now().UnixNano()))
+// Build a new Resolver to route to the given target
+func (b *JSONGateResolverBuilder) Build(target resolver.Target, cc resolver.ClientConn, opts resolver.BuildOptions) (resolver.Resolver, error) {
+	attrs := target.URL.Query()
 
-	// Immediately load the initial config
-	addrs, hash, err := r.resolve()
-	if err == nil {
-		// if we parse ok, populate the local address store
-		r.cc.UpdateState(resolver.State{Addresses: *addrs})
+	// If the config specifies a pool type attribute, then the caller must supply it in the connection
+	// attributes, otherwise reject the request.
+	poolType := ""
+	if b.poolTypeField != "" {
+		poolType = attrs.Get(b.poolTypeField)
+		if poolType == "" {
+			return nil, fmt.Errorf("pool type attribute %s not in target", b.poolTypeField)
+		}
 	}
 
-	// Start a config watcher
-	r.ticker = time.NewTicker(100 * time.Millisecond)
-	fileStat, err := os.Stat(r.jsonPath)
-	if err != nil {
-		return
+	// Affinity on the other hand is just an optimization
+	affinity := ""
+	if b.affinityField != "" {
+		affinity = attrs.Get(b.affinityField)
 	}
-	go func() {
-		for range r.ticker.C {
-			checkFileStat, err := os.Stat(r.jsonPath)
-			if err != nil {
-				log.Errorf("Error stat'ing config %v\n", err)
-				continue
-			}
-			isUnchanged := checkFileStat.Size() == fileStat.Size() || checkFileStat.ModTime() == fileStat.ModTime()
-			if isUnchanged {
-				// no change
-				continue
-			}
-
-			fileStat = checkFileStat
-			log.V(100).Infof("Detected config change\n")
 
-			addrs, newHash, err := r.resolve()
-			if err != nil {
-				// better luck next loop
-				// TODO: log this
-				log.Errorf("Error resolving config: %v\n", err)
-				continue
-			}
+	log.V(100).Infof("Start discovery for target %v poolType %s affinity %s\n", target.URL.String(), poolType, affinity)
 
-			// Make sure this wasn't a spurious change by checking the hash
-			if bytes.Equal(hash, newHash) && newHash != nil {
-				log.V(100).Infof("No content changed in discovery file... ignoring\n")
-				continue
-			}
+	r := &JSONGateResolver{
+		target:     target,
+		clientConn: cc,
+		poolType:   poolType,
+		affinity:   affinity,
+	}
 
-			hash = newHash
+	b.update(r)
+	b.resolvers = append(b.resolvers, r)
```

Comment:
The reason we never clear this is because we expect to create exactly one resolver per target during the lifetime of the process, right?

Reply:
Correct. We either reuse the existing resolver or create a new one, we never need to remove them.